### ISSUE TITLE
Accessing symlinked Library Datasets whose source file no longer exist produces inconsistent errors 

### ISSUE THREAD

**Issue by luke-c-sargent** [galaxyproject/galaxy#8578]
Hello,

# TL;DR:  

Symlinked files that no longer exist cause inconsistent errors in places where they are de-referenced from Dataset abstraction to actual file; usually this is a cryptic message that takes up the whole screen, superseding Galaxy's UI, but in at least one instance this leaks the local path of the file.  

---

I am encountering inconsistent behavior with symlinked Library datasets local to the Galaxy server (or at least presented as such via network mount). When such a file is deleted or not mounted or otherwise unavailable, the Dataset abstraction still exists, so many operations on this 'file' succeed, only to fail at whatever point resolution of the Dataset to the actual, underlying data is requested.

To reproduce:
- create library
- add local file using the symlink capacity
- delete local file

It is partially visible in the library via the stored peek, and can be added to a history.

## Errors
When trying to download from the library, it produces an error (which is the only content in the window; the Galaxy UI is no longer present):
![Screen Shot 2019-09-05 at 11 54 02 AM](https://user-images.githubusercontent.com/11185084/64371538-a1bccf00-cfd5-11e9-89b6-00ec3bca5b6d.png)
Log: 
```
Traceback (most recent call last):
  File "lib/galaxy/webapps/galaxy/api/library_datasets.py", line 673, in download
    archive.add(ldda.dataset.file_name, path, check_file=True)
  File "lib/galaxy/util/streamball.py", line 28, in add
    raise ObjectNotFound
ObjectNotFound: No such object found.
```
When trying to view with the eye icon:
![Screen Shot 2019-09-05 at 11 52 09 AM](https://user-images.githubusercontent.com/11185084/64371789-4dfeb580-cfd6-11e9-97ed-7968310601ff.png)

Log:
```
127.0.0.1 - - [05/Sep/2019:12:10:01 -0700] "GET /datasets/f597429621d6eb2b/display/?preview=True HTTP/1.1" 404 181 "http://127.0.0.1:8080/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:68.0) Gecko/20100101 Firefox/68.0"
```

When trying to download via the history floppy disk icon:
(No screenshot, just goes to a blank screen without the Galaxy UI)
Log:
```
Traceback (most recent call last):
  File "lib/galaxy/web/framework/middleware/batch.py", line 80, in __call__
    return self.application(environ, start_response)
  File "lib/galaxy/web/framework/middleware/request_id.py", line 15, in __call__
    return self.app(environ, start_response)
  File "lib/galaxy/web/framework/middleware/xforwardedhost.py", line 23, in __call__
    return self.app(environ, start_response)
  File "lib/galaxy/web/framework/middleware/translogger.py", line 71, in __call__
    return self.application(environ, replacement_start_response)
  File "lib/galaxy/web/framework/middleware/error.py", line 166, in __call__
    response = self.exception_handler(exc_info, environ)
  File "lib/galaxy/web/framework/middleware/error.py", line 181, in exception_handler
    get_vars = wsgilib.parse_querystring(environ)
  File "/Users/sargentl/code/temp/galaxy_history_mods/.venv/lib/python2.7/site-packages/paste/request.py", line 107, in parse_querystring
    strict_parsing=False)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urlparse.py", line 409, in parse_qsl
    pairs = [s2 for s1 in qs.split('&') for s2 in s1.split(';')]
  File "lib/galaxy/web/framework/middleware/error.py", line 154, in __call__
    app_iter = self.application(environ, sr_checker)
  File "/Users/sargentl/code/temp/galaxy_history_mods/.venv/lib/python2.7/site-packages/paste/recursive.py", line 85, in __call__
    return self.application(environ, start_response)
  File "/Users/sargentl/code/temp/galaxy_history_mods/.venv/lib/python2.7/site-packages/paste/httpexceptions.py", line 640, in __call__
    return self.application(environ, start_response)
  File "lib/galaxy/web/framework/base.py", line 143, in __call__
    return self.handle_request(environ, start_response)
  File "lib/galaxy/web/framework/base.py", line 222, in handle_request
    body = method(trans, **kwargs)
  File "lib/galaxy/webapps/galaxy/controllers/dataset.py", line 251, in display
    return data.datatype.display_data(trans, data, preview, filename, to_ext, offset=offset, ck_size=ck_size, **kwd)
  File "lib/galaxy/datatypes/data.py", line 449, in display_data
    trans.response.headers['Content-Length'] = int(os.stat(data.file_name).st_size)
OSError: [Errno 2] No such file or directory: '/tmp/bork'
```

All of this is reasonable; can't view / download something that doesn't exist. 

**A question**: does this necessitate action, and what sort? This is something an admin would have to deal with, but it would probably be nice to tell the user something's up instead of blank screens or leaking the local path of the file. Is there a convenient place to put this checking logic shared by all Dataset file access? The errors crop up in various places in the code base (generally where `os` is invoked), but my hope is there is some step before all of these invocations that can incorporate a check (only symlinked files suffer from this, so we could limit checking to files with 'external_filename' set).

A separate but related question: should we take any other steps besides warning? So many different scenarios can give rise to this issue that it may not always be a good idea to simply prune this file from the Library and move on with your life.
